sound/soc/sof/intel/hda-dsp.c

```diff
@@ -381,6 +406,100 @@ int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
 	return ret;
 }
 
+static int hda_dsp_set_D3_state(struct snd_sof_dev *sdev,
+				const struct sof_dsp_power_state *target_state)
+{
+	/*
+	 * Sanity check for illegal state transitions
+	 * The only allowed transition is: D0I0 -> D3
+	 */
+	switch (sdev->dsp_power_state.state) {
+	case SOF_DSP_D0:
+		if (sdev->dsp_power_state.substate != SOF_HDA_DSP_D0I0) {
+			dev_err(sdev->dev,
+				"error: transition from D0I3 -> D3 not allowed\n");
+			return -EINVAL;
+		}
+		break;
+	case SOF_DSP_D3:
+		/* already in D3 */
+		return 0;
+	default:
+		dev_err(sdev->dev, "error: transition from %d to %d not allowed\n",
+			sdev->dsp_power_state.state, target_state->state);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/* All DSP power state transitions are initiated by the driver */
+int hda_dsp_set_power_state(struct snd_sof_dev *sdev,
+			    const struct sof_dsp_power_state *target_state)
+{
+	int ret = 0;
+
+	/* Nothing to do if the DSP is already in the requested state */
+	if (target_state->state == sdev->dsp_power_state.state &&
+	    target_state->substate == sdev->dsp_power_state.substate)
+		return 0;
+
+	switch (target_state->state) {
+	case SOF_DSP_D0:
+		ret = hda_dsp_set_D0_state(sdev, target_state);
+		break;
+	case SOF_DSP_D3:
+		/* Follow the regular flow for DSP D3 */
+		ret = hda_dsp_set_D3_state(sdev, target_state);
+		break;
+	default:
+		dev_err(sdev->dev, "error: target state unsupported %d\n",
+			target_state->state);
+		return -EINVAL;
+	}
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"failed to set requested target DSP state %d substate %d\n",
+			target_state->state, target_state->substate);
+		return ret;
+	}
+
+	sdev->dsp_power_state = *target_state;
+	return ret;
+}
+
+/*
+ * Audio DSP states may transform as below:-
+ *
```

Comment:
why do you need to have the two arcs on the right side " D0I3 compatible stream opened only, timeout", ow is this different from "D0i3 suspend". Or did you mean "S0ix suspend" instead of "D0i3 suspend" if you want to represent the possible transitions (opportunistic and forced by a suspend)?